File: app/schemas/budget.py
import uuid
from pydantic import BaseModel


class BudgetCalcRequest(BaseModel):
    project_id: uuid.UUID | None = None
    city_code: str | None = None
    city_name: str | None = None
    inner_area: float
    layout_type: str
    tier: str = 'standard'
    floor_preference: str = 'tile'
    bathroom_count: int = 1
    special_needs: list[str] = []


class BudgetItemOut(BaseModel):
    id: uuid.UUID
    scheme_id: uuid.UUID | None = None
    room_id: uuid.UUID | None = None
    standard_item_id: uuid.UUID | None = None
    category: str
    item_name: str
    pricing_mode: str | None = None
    quantity: float | None = None
    unit: str | None = None
    material_unit_price: float | None = None
    labor_unit_price: float | None = None
    accessory_unit_price: float | None = None
    loss_rate: float | None = None
    subtotal: float | None = None
    is_user_modified: bool
    data_source: str | None = None
    remark: str | None = None
    sort_order: int | None = None

    model_config = {'from_attributes': True}


class BudgetSchemeOut(BaseModel):
    id: uuid.UUID
    project_id: uuid.UUID | None = None
    tier: str
    total_amount: float | None = None
    material_amount: float | None = None
    labor_amount: float | None = None
    accessory_amount: float | None = None
    management_fee: float | None = None
    contingency: float | None = None
    ai_explanation: str | None = None
    items: list[BudgetItemOut] = []

    model_config = {'from_attributes': True}


class BudgetResultOut(BaseModel):
    project_id: uuid.UUID
    schemes: list[BudgetSchemeOut]
    missing_items: list[dict] = []
    suggestions: list[str] = []

File: src/types/budget.ts
import { TierLevel } from './house';

export type PricingMode = 'area' | 'linear_meter' | 'point' | 'quantity';

export interface BudgetCategory {
    id: string;
    name: string;
    nameZh: string;
    icon: string;
    order: number;
}

export interface BudgetItem {
    id: string;
    scheme_id: string;
    room_id: string | null;
    category: string;
    standard_item_id: string;
    item_name: string;
    pricing_mode: PricingMode;
    quantity: number;
    unit: string;
    material_unit_price: number;
    labor_unit_price: number;
    accessory_unit_price: number;
    loss_rate: number;
    subtotal: number;
    data_source: string;
    is_user_modified: boolean;
    remark?: string;
}

export interface BudgetScheme {
    id: string;
    project_id: string;
    tier: TierLevel;
    total_amount: number;
    material_amount: number;
    labor_amount: number;
    management_fee: number;
    contingency: number;
    items: BudgetItem[];
    createdAt: string;
    // Legacy mapping (to be removed after refactoring components)
    totalBudget?: number;
}

export interface BudgetResult {
    project_id: string;
    schemes: BudgetScheme[];
    economy?: BudgetScheme;
    standard?: BudgetScheme;
    premium?: BudgetScheme;
    missing_items: any[];
    suggestions: string[];
    missingItems: MissingItem[]; // Legacy support
    optimizations: OptimizationSuggestion[]; // Legacy support
    overBudget: boolean;
    overBudgetAmount: number;
}

export interface MissingItem {
    id: string;
    ruleName: string;
    itemName: string;
    riskLevel: 'high' | 'medium' | 'low';
    estimatedPriceMin: number;
    estimatedPriceMax: number;
    explanation: string;
    askTemplate: string;
    shouldInclude: boolean;
}

export interface OptimizationSuggestion {
    id: string;
    type: 'save' | 'must_keep' | 'alternative';
    category: string;
    title: string;
    description: string;
    savingAmount?: number;
}
